### ISSUE TITLE
Experiment with a Teaclave backend

### ISSUE THREAD

**Issue by blaxill** [project-oak/oak#424]
The Oak project may decide to move away from Asylo and to a Rust native framework. I'm going to experiment with [Teaclave](https://github.com/apache/incubator-teaclave-sgx-sdk/) by writing a skeleton oak runtime for it.

The main objectives are to:

  - Evaluate teaclave as an enclave platform
  - Discover what if any changes to oak runtime may be beneficial for other enclave platform support.

**Comment by tiziano88** [comment]
@blaxill could you summarize the findings here, and consider closing this issue?

**Comment by blaxill** [comment]
- [Enclave platforms and Rust in Oak (Summary)](https://github.com/project-oak/oak/issues/471)

Teaclave and other options provide `std` libraries trying to achieve compatibility with the Rust standard library. Unfortunately none of the options are a perfect replacement, requiring dependency modifications when compiling packages. Teaclave generally requires more patching than e.g. Fortanix, but both suffer when compiling crates that have dependencies on e.g. `net2` which wraps more detailed socket operations provided by the OS. It's unsurprising that the enclave frameworks don't work with OS wrapping crates, but also kills the dream of building things in `std` without modification.

The current plan is to continue developing a `no_std` rust based runtime that works with the Asylo framework, and then in the future look to make the runtime enclave agnostic.
